File: python/stringify_num.py
import sys

# Number lookup tables.

_number_digits: list = [
    "zero", "one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine"
]

_number_prefixes: list = [
    "twen", "thir", "four", "fif", "six", "seven", "eigh", "nine"
]

_hundred_suffixes: list = [
    "thousand", "million", "billion", "trillion", "quadrillion",
    "quintillion", "sextillion", "septillion", "octillion", "nonillion",
    "undecillion", "duodecillion", "tredecillion", "quattuordecillion", "quindecillion",
    "sexdecillion", "septendecillion", "octodecillion", "novemdecillion", "vigintillion"
]

# Convert a digit to 0 through 9.
def string_digit(digit: int) -> str:
    if digit >= 0 and digit <= 9:
        return _number_digits[digit]
    else:
        raise Exception(str(digit) + " is out of digit range (0-9)")
    
# Convert a teen to 10 through 19.
def _string_teen(teen: int) -> str:
    if teen == 0: return "ten"
    elif teen == 1: return "eleven"
    elif teen == 2: return "twelve"
    elif teen >= 3 and teen <= 9:
        return _number_prefixes[teen - 2] + "teen"
    else:
        raise Exception(str(teen) + " is out of teen range (10-19)")

# Convert a ten to 20 through 90.
def _string_ten(ten: int) -> str:
    if ten == 0:
        return ""
    if ten >= 2 and ten <= 9:
        return _number_prefixes[ten - 2] + "ty"
    else:
        raise Exception(str(ten) + " is out of tens range (20-90)")

def _stringify_hundred(number: int, insert_and: bool) -> str:
    if number >= 1000:
        raise Exception("number is greater than 3 digits!")

    result: str = ""
    # For an exact hundreds value, discard the previous number
    if number >= 100 and number % 100 == 0:
        result = string_digit(number // 100 % 10) + " hundred"
    else: 
        # Check for the teen range and handle it specially.
        if number % 100 < 20 and number % 100 > 9:
            result = _string_teen(number % 10)
        else:
            # Otherwise construct the first two digits.
            if number % 100 < 10:
                result = string_digit(number % 10)
            else:
                result = _string_ten(number // 10 % 10)
                if number % 10 != 0:
                    result += " " + string_digit(number % 10)
        # Now check hundreds
        if number >= 100:
            result = string_digit(number // 100 % 10) + " hundred " + \
                ("and " if insert_and else "") + result
    return result

def stringify_number(number: int, insert_and: bool = False) -> str:
    if number >= 10**64:
        raise Exception("Maximum number is 999 vigintillion (63 digits)")
    
    negative: bool = number < 0
    number = abs(number)
    result: str = _stringify_hundred(number % 1000, insert_and)

    factor: int = 1000
    factor_index: int = 0
    while number >= factor:
        if number % factor == 0:
            result = ""
        result = _stringify_hundred(number // factor % 1000, insert_and) + \
            " " + _hundred_suffixes[factor_index] + \
            ((" " + result) if result else "")
        factor *= 1000
        factor_index += 1

    if negative:
        result = "negative " + result
    return result

def __main__():
    try:
        if len(sys.argv) > 1:
            input_number: int = int(sys.argv[1])
        else:
            input_number: int = int(input("Enter a number: "))
    except:
        print("Invalid Input! Please enter a single number without any commas.")
        sys.exit()

    print("Number " + str(input_number) + " is " + stringify_number(input_number))

if __name__ == "__main__":
    __main__()

File: cplusplus/stringify-num/src/stringify-num.cpp
#include <array>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <string>

std::array<std::string, 10> number_digits = {
    "zero", "one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine"
};

std::array<std::string, 8> number_prefixes = {
    "twen", "thir", "four", "fif", "six", "seven", "eigh", "nine"
};

// All possible suffixes for a signed 64-bit integer.
std::array<std::string, 6> hundred_suffixes = {
    "thousand", "million", "billion", "trillion", "quadrillion", "quintillion"
};

std::string stringify_hundred(uint16_t number, bool insert_and) {
	std::string result;

	// Specially handle numbers under 10 (specifically for zero).
	if (number < 10) {
		return number_digits[number];
	}
	
	// Add hundreds
	if (number >= 100) {
		result = number_digits[number / 100 % 10] + " hundred";
		// If there are no remaining digits, exit early.
		if (number % 100 == 0) {
			return result;
		} else {
			result += " ";
			result += insert_and ? "and " : "";
		}
	}
	// Check for the teen range and handle it specially.
	if (number % 100 < 20 and number % 100 > 9) {
		switch (number % 10) {
		case 0: result + "ten";
		case 1: result + "eleven";
		case 2: result + "twelve";
		default:
			return result + number_prefixes[number % 10 - 2] + "teen";
		}
	} else if (number % 100 >= 20) {
		result += number_prefixes[number / 10 % 10 - 2] + "ty";
		if (number % 10 == 0) {
			return result;
		} else {
			result += " ";
		}
	}
	return result + number_digits[number % 10];
}

std::string stringify_number(int64_t number, bool insert_and = false) {
	bool negative = number < 0;
	number = std::abs(number);
	std::string result = stringify_hundred(number % 1000, insert_and);

	int64_t factor = 1000;
	uint8_t factor_index = 0;
	while (number >= factor) {
		if (number / factor % 1000 != 0) {
			result = stringify_hundred(number / factor % 1000, insert_and) + \
				" " + hundred_suffixes[factor_index] + ((number % factor) ? (" " + result) : "");
		}
		factor *= 1000;
		factor_index += 1;
	}

	if (negative) {
		return "negative " + result;
	}
	return result;
}
